rewrite feedback

gaborbernat opened the pull request:
This is a PR meant to allow people to give feedback on the rewrite. For such big rewrite, I'm not sure Github UI is good so I'd recommend checking out the rewrite branch. Opened the PR though to allow people to comment in the code inline - or attach their feedback as a comment. If you take the time and give feedback I'd like to say thank you very much, appreciated 🙇 

A bit of context on the design, the core rewrite idea is to make the project as fast as possible and relatively easy to extend/maintain. We now have a config file support, which allows users to choose their own defaults globally (``virtualenv -h`` now shows where defaults come from, experience prooved that's there's no such thing as good default for everyone, and these can be also used to globally fix when some dependency pip/setuptools breaks the world). Furthermore, we still support python 2; and plan to support it for the foreseeable future - pypy is still supported, and we plan to support that. As such all code must support python 2.7+. That being said, the rewrite simplifies the python 2 env creation, to something that looks a lot more like python 3 style via a pyenv.cfg. 

The idea is also to support a clean programmatic API for tox/pipenv/etc (e.g. what's the site-package folder for this env, without creating it; reuse environment discovery, etc).

The project is split into four main sections:
- interpreter discovery (when the user passes in ``-p 3.8-64`` discover/find an interpreter that satisfies those requirements - supports PEP-514 or path discovery mechanism)
- interpreter creation (create the actual environment; either via delegation to venv, or ourselves - here we also implement ourselves creation code even when venv is available to avoid process creation overhead 🤷‍♂ which can be slow on some platforms especially)
- seed packages (install pip/setuptools - either the old way via pip, or a faster method via app data image - 95% of the environment creation is spent within this phase, pip is slow mostly because has to support a lot of features, the app data image installation works around this by needing to support a smaller subset of pip - with this 200ms creation time is now possible)
- activation scripts (now supports unicode paths, warns when a path would break, etc. - also now users can add plugins to add their own activation scripts)

CC @pfmoore @pradyunsg @RonnyPfannschmidt 

PS: There still are some release blockers beyond this review -> https://github.com/pypa/virtualenv/milestone/7

pfmoore commented:
Some initial (very simple) tests on Windows look OK.

* Installing the new virtualenv in a venv works fine (which is good, as it makes it easier to test without affecting your system Python).
* Similarly, the new virtualenv can be installed using pipx (good for providing a single central installation).
* Using `-p 3.8` resulted in a warning about invalid PEP 514 data in my registry for Python 2.7. No big deal as it's a warning, but as it's for a Python version I wasn't trying to use, the warning is unnecessary (and potentially scary for many users). I don't know how the data that was present - I no longer have Python 2.7 installed, although I did at one point. Maybe installing and uninstalling leaves it?

I tried to test with Python 2.7, using a custom Python 2.7 install (installed via Scoop) and `virtualenv -p C:\Users\xxx\scoop\apps\python-2.7.17\1.0\Python-2.7.17\python.exe xx`. The resulting environment didn't include a copy of `python27.dll`, and so `python.exe` failed. Virtualenv 16.7.9 doesn't have this issue with this Python installation. (I don't have anywhere I can install the "official" Python 2.7 at the moment, so I can't test this any further).

By the way, the new virtualenv seems to install `pydoc.bat`. The old one didn't - ideally I'd prefer this *not* to be installed, as it interferes with command completion (`myenv\Scripts\py<TAB>`). I realise this is a minor use case, but [xkcd is on my side](https://xkcd.com/1172/) 🙂 

But overall, this looks **very** nice. Good job!

asottile commented on src/virtualenv/interpreters/discovery/py_info.py:
```diff
@@ -0,0 +1,256 @@
+    def _find_possible_exe_names(self):
+        name_candidate = OrderedDict()
+        for name in [self.implementation, "python"]:
+            for at in range(3, -1, -1):
```
```suggestion
            for at in (3, 2, 1, 0):
```

slightly more readable -- but now that I've expanded this i'm confused what it's trying to do? do we expect to see `python0` executables? is this going to break if/when python4?

gaborbernat replied:
the at is used to count the number of version parts to use; e.g. ``python3.8.1, python3.8, python3, python``

asottile commented on src/virtualenv/interpreters/discovery/py_info.py:
```diff
@@ -0,0 +1,256 @@
+
+    @classmethod
+    def _load_for_exe(cls, exe):
+        from virtualenv.util.subprocess import subprocess, Popen
```
hmmm, this seems to be non-stdlib imports -- is this OK (given the docstring)

gaborbernat replied:
the from_exe method is an exception as it's only used from within virtualenv 👍

asottile commented on src/virtualenv/interpreters/discovery/py_spec.py:
```diff
@@ -0,0 +1,115 @@
+
+    @classmethod
+    def from_string_spec(cls, string_spec):
+        impl, major, minor, patch, arch, path = None, None, None, None, None, None
```
```suggestion
        impl = major = minor = patch = arch = path = None
```

slightly less counting necessary when adding / removing a variable ;)

asottile commented on src/virtualenv/run.py:
```diff
@@ -0,0 +1,190 @@
+    ]
+    [next(elem) for elem in elements]  # add choice of types
+    parser.parse_known_args(args, namespace=options)
+    [next(elem) for elem in elements]  # add type flags
```
these list comprehensions seem very suspicious -- any way this can be done that's more explicit?

asottile commented on src/virtualenv/run.py:
```diff
@@ -0,0 +1,190 @@
+    yield
+    seeder_class = seeder_types["none" if options.without_pip is True else options.seeder]
+    seeder_class.add_parser_arguments(seed_parser)
+    yield
```
ah yes I see this now -- I'd maybe write this as three functions you call in order -- the generator trick seems way too clever (from a maintainability perspective)

asottile commented on src/virtualenv/seed/embed/wheels/__init__.py:
```diff
@@ -0,0 +1,12 @@
+from __future__ import absolute_import, unicode_literals
+
+BUNDLE_SUPPORT = {
+    "3.9": {"pip": "pip-19.3.1-py2.py3-none-any.whl", "setuptools": "setuptools-42.0.2-py2.py3-none-any.whl"},
```
this seems to imply that newer versions will require a  change in `virtualenv` -- should it (does it already?) default to the latest?

(I had mostly fixed this in `tox` to not need updates every new version, pyflakes is growing some code for this too)

gaborbernat replied:
newer versions automatically use MAX (see few lines below); we still check for wheel compatibility so it's alright

asottile commented on src/virtualenv/util/__init__.py:
```diff
@@ -0,0 +1,10 @@
+    import ConfigParser
+except ImportError:
+    # noinspection PyPep8Naming
+    import configparser as ConfigParser
```
I tend to prefer being explicit here:

```
if sys.verison_info[0] == 2:
    import ConfigParser
else:
    import configparser as ConfigParser
```

this can prevent a class of problems where a nasty third party package provides a `ConfigParser` file in python3

asottile commented on src/virtualenv/util/path/_sync.py:
```diff
@@ -0,0 +1,50 @@
+
+import six
+
+HAS_SYMLINK = hasattr(os, "symlink")
```
modern python3s on windows always have this attribute, just it's sometimes nonfunction (not sure it matters here)

gaborbernat replied:
yeah, I know but they're not reliable; tested them a bit and even in my case had mixed results, so we don't use it for windows yet

asottile commented on tasks/make_zipapp.py:
```diff
@@ -17,16 +17,17 @@ def main():
-        filenames = ["LICENSE.txt", "virtualenv.py"]
-        for whl in os.listdir(os.path.join(args.root, "virtualenv_support")):
-            filenames.append(os.path.join("virtualenv_support", whl))
+    filenames = {"LICENSE.txt": "LICENSE.txt", os.path.join("src", "virtualenv.py"): "virtualenv.py"}
```
I suspect this zipapp is nonfunctional -- the tests are probably only passing because `virtualenv`'s dependencies are installed in those environments

gaborbernat replied:
this is one of those tbd https://github.com/pypa/virtualenv/issues/1464 things

asottile commented on src/virtualenv/seed/embed/base_embed.py:
```diff
@@ -0,0 +1,81 @@
+        for extra, package in [
+            ("", "pip"),
+            ("", "setuptools"),
+            ("N/A - kept only for backwards compatibility; ", "wheel"),
```
I suspect the removal of `wheel` from virtualenvs will break more people than it should -- it was certainly an issue for us when wheel was introduced, I imagine the removal will be similarly annoying/painful

gaborbernat replied:
The removal is there to mirror what venv does... venv must have had a good reason why to remove wheel 🤷‍♂ ?

asottile replied:
hmmmm maybe I'm misremembering the history here but I think `virtualenv` added `wheel` *after* `venv` existed -- and maybe it was just not added to `venv`?

gaborbernat replied:
either way if venv can get away with not having it, and virtualenv now want to be feature parity with venv seems like a good idea to not have it 😊 unless you have use cases where not having wheel seeded breaks the world

asottile replied:
it makes default cache performance of `pip` way way worse (since it can't prebuild wheels and put them into cache) -- I'd rather we keep it

gaborbernat replied:
@pfmoore as pip maintainer can you confirm that under venv build performance is significantly worse due to not bundling wheel? (if so I wonder why no one opened an upstream bug for it)

asottile replied:
for example:

```bash
#!/usr/bin/env bash
set -euxo pipefail

clean_env() {
    rm -rf ~/.cache/pip
    rm -rf venv
    virtualenv venv --no-download >& /dev/null
}

run_test() {
    venv/bin/pip install -qq pyyaml
    venv/bin/pip uninstall -qqy pyyaml
    time venv/bin/pip install -qq pyyaml
}


: with wheel
clean_env
run_test

: without wheel
clean_env
venv/bin/pip uninstall -qqy wheel
run_test
```


```console
$ bash t.sh 
+ : with wheel
+ clean_env
+ rm -rf /home/asottile/.cache/pip
+ rm -rf venv
+ virtualenv venv --no-download
+ run_test
+ venv/bin/pip install -qq pyyaml
+ venv/bin/pip uninstall -qqy pyyaml
+ venv/bin/pip install -qq pyyaml

real	0m0.900s
user	0m0.726s
sys	0m0.053s
+ : without wheel
+ clean_env
+ rm -rf /home/asottile/.cache/pip
+ rm -rf venv
+ virtualenv venv --no-download
+ venv/bin/pip uninstall -qqy wheel
+ run_test
+ venv/bin/pip install -qq pyyaml
+ venv/bin/pip uninstall -qqy pyyaml
+ venv/bin/pip install -qq pyyaml

real	0m9.808s
user	0m9.333s
sys	0m0.249s
```

without `wheel` installed, `pip` cannot build a wheel (and therefore can't cache the built output)

gaborbernat replied:
oh, I'll add it back then, but puzzled how venv got away with not having it for the last 5 years 🤦‍♂

gaborbernat replied:
addressed by https://github.com/pypa/virtualenv/issues/1486

gaborbernat commented on setup.cfg:
```diff
@@ -1,81 +1,108 @@
+    appdirs >= 1.4.3
+    entrypoints >= 0.3, <1
+    pathlib2 >= 2.3.3, < 3; python_version < '3.4' and sys.platform != 'win32'
+    distlib >= 0.3.0, <1; sys.platform == 'win32'
```
the expectations is that for where people want to bootstrap they should use zipapp; please detail use cases where that doesn't solve your problem 🤷‍♂

asottile replied:
with the rewrite branch the zipapp currently doesn't work (and as far as I can tell, the zipapp doesn't handle the dependencies either)

gaborbernat replied:
that's tbd via https://github.com/pypa/virtualenv/issues/1464

gaborbernat commented on src/virtualenv/activation/bash/activate.sh:
```diff
@@ -46,7 +46,7 @@ deactivate () {
 deactivate nondestructive
 
-VIRTUAL_ENV="__VIRTUAL_ENV__"
+VIRTUAL_ENV='__VIRTUAL_ENV__'
```
the ``'`` here allows having ``$`` within the path... this is a shell dependent feature, hence why the quote style is specified here; the quotes are never replaced just the ``__something part__``

asottile replied:
right, but now it can't work with single quotes within the path -- using `pipes.quote` when substituting will make both of them work here

gaborbernat replied:
tracked under https://github.com/pypa/virtualenv/issues/1488

gaborbernat commented on src/virtualenv/activation/via_template.py:
```diff
@@ -0,0 +1,38 @@
+    def replacements(self, creator, dest_folder):
+        return {
+            "__VIRTUAL_PROMPT__": "" if self.flag_prompt is None else self.flag_prompt,
+            "__VIRTUAL_ENV__": six.ensure_text(str(creator.dest_dir)),
```
six.text_type uses the default encoding which in a lot of cases will break; I've tried going that route and ran into lots of issues; when we'll drop python 2 (not anytime soon as pypy will continue to be supported) we'll replace these instances manually, that's okay

asottile replied:
doesn't `str(...)` also go through default encoding?

gaborbernat replied:
only if the argument to it is a string or text... otherwise it's the objects __str__ ... which may choose its own encoding

asottile replied:
the same is true for `six.text_type` -- that'll go through `__unicode__` in python2 and `__str__` in python3 -- encodings don't get involved at all with that

this is why I suggested it over the current code, which has to go through a full encode / decode (and potentially encoding / decoding errors)

gaborbernat commented on src/virtualenv/activation/via_template.py:
```diff
@@ -0,0 +1,38 @@
+
+    def _generate(self, replacements, templates, to_folder):
+        for template in templates:
+            text = pkgutil.get_data(self.__module__, str(template)).decode("utf-8")
```
Can you qualify on this optional aspect?

asottile replied:
`get_data` is part of the optional PEP 302 api: https://www.python.org/dev/peps/pep-0302/#optional-extensions-to-the-importer-protocol

gaborbernat replied:
so we could go along the way of importlib.resources... but that means another dependency 🤷‍♂

asottile replied:
only in python<3.7 -- it's standard library after that :)

gaborbernat replied:
done

gaborbernat commented on src/virtualenv/config/env_var.py:
```diff
@@ -0,0 +1,27 @@
+    :return:
+    """
+    environ_key = "VIRTUALENV_{}".format(key.upper())
+    if environ_key in os.environ:
```
you have a typo here 🤔

asottile replied:
lol dangit, fixed :)

gaborbernat commented on src/virtualenv/info.py:
```diff
@@ -0,0 +1,26 @@
+from virtualenv.util.path import Path
+
+IS_PYPY = hasattr(sys, "pypy_version_info")
+PY3 = sys.version_info[0] == 3
```
Guido spoke up that we'll continue having 3.10 instead of 4.x, and until there's a clear signal from the steering council that 4.x will not be a breaking change I consider this to be the correct way of checking for Python 3 compatibility 👍

asottile replied:
Even still, this code is likely to live much longer than normal code -- I'd strongly suggest using `>=` here for future proofing

It is very unlikely that python4 will be breaking in the same manner as python3 was -- guido has also spoken in the past that when python4 comes it will be "just another python3"

The main reason for preferring python3.10 is because people write incorrect code like `== 3` which prevents the ecosystem from migrating

gaborbernat commented on src/virtualenv/interpreters/create/cpython/cpython3.py:
```diff
@@ -0,0 +1,36 @@
+
+
+class CPython3Windows(CPythonWindows, CPython3):
+    """"""
```
yes 👍 will go through the code and document throughout once I get the thing in a stable place

gaborbernat commented on src/virtualenv/interpreters/create/creator.py:
```diff
@@ -0,0 +1,182 @@
+            refused = set(raw_value) - {
+                c
+                for c, i in ((char, char.encode(encoding)) for char in raw_value)
+                if c == "?" or i != six.ensure_str("?")
```
I'll keep the as-is format for consistency with six.ensure_text 👍

gaborbernat commented on src/virtualenv/interpreters/create/creator.py:
```diff
@@ -0,0 +1,182 @@
+            )
+        if os.pathsep in raw_value:
+            raise ArgumentTypeError(
+                "destination {!r} must not contain the path separator ({}) as this would break "
```
the path separator is ``:`` not ``/`` that's ``os.sep``

asottile replied:
oh right 🤦‍♂ -- I always get confused by these names

gaborbernat commented on src/virtualenv/interpreters/create/creator.py:
```diff
@@ -0,0 +1,182 @@
+            raise ArgumentTypeError(
+                "the file system codec ({}) does not support characters {!r}".format(encoding, refused)
+            )
+        if os.pathsep in raw_value:
```
good point

gaborbernat replied:
done now

gaborbernat commented on src/virtualenv/interpreters/create/creator.py:
```diff
@@ -0,0 +1,182 @@
+    def run(self):
+        if self.dest_dir.exists() and self.clear:
+            logging.debug("delete %s", self.dest_dir)
+            shutil.rmtree(str(self.dest_dir), ignore_errors=True)
```
🤔 haven't thought of that edge case, seems like a feature request that deserves its own test 👍 and PR

gaborbernat commented on src/virtualenv/interpreters/create/creator.py:
```diff
@@ -0,0 +1,182 @@
+
+
+def get_env_debug_info(env_exe, debug_script):
+    cmd = [six.ensure_text(str(env_exe)), six.ensure_text(str(debug_script))]
```
Oh I know, but added the workaround here https://github.com/pypa/virtualenv/blob/rewrite/src/virtualenv/util/subprocess/_win_subprocess.py

gaborbernat commented on src/virtualenv/interpreters/create/venv.py:
```diff
@@ -0,0 +1,72 @@
+    def lib_dir(self):
+        base = self.dest_dir / ("Lib" if self.interpreter.os == "nt" else "lib")
+        if self.interpreter.os != "nt":
+            base = base / self.interpreter.python_name
```
``pypy`` support to be added 👍 I'm just working on that, but that's a separate PR

gaborbernat commented on src/virtualenv/interpreters/discovery/builtin.py:
```diff
@@ -0,0 +1,135 @@
+def check_path(candidate, path):
+    _, ext = os.path.splitext(candidate)
+    if sys.platform == "win32" and ext != ".exe":
+        candidate = candidate + ".exe"
```
I've tried to remove all features for which I don't have a well-defined use-case and test 🤔 so features we have are actually tested 👍 so what you're saying here seems like a valid feature request, but would need a concrete use case for it

gaborbernat commented on src/virtualenv/interpreters/discovery/py_info.py:
```diff
@@ -0,0 +1,256 @@
+"""
+The PythonInfo contains information about a concrete instance of a Python interpreter
+
+Note: this file is also used to query target interpreters, so can only use standard library methods
```
this file is tested thoroughly and via review; I prefer to put it where it belongs rather than stdlib only package 🤷‍♂

gaborbernat commented on src/virtualenv/interpreters/discovery/py_info.py:
```diff
@@ -0,0 +1,256 @@
+        self.exec_prefix = getattr(sys, "exec_prefix", None)
+
+        try:
+            __import__("venv")
```
venv we use to create the virtualenv, not to seed it; your failure there occurs during the package seeding phase that we never invoke 👍 (we always use our own seeding logic)

gaborbernat commented:
> Using -p 3.8 resulted in a warning about invalid PEP 514 data in my registry for Python 2.7. No big deal as it's a warning, but as it's for a Python version I wasn't trying to use, the warning is unnecessary (and potentially scary for many users).

This seems a good idea, will try to add.

> I tried to test with Python 2.7, using a custom Python 2.7 install (installed via Scoop) and virtualenv -p C:\Users\xxx\scoop\apps\python-2.7.17\1.0\Python-2.7.17\python.exe xx. The resulting environment didn't include a copy of python27.dll, and so python.exe failed.

Interesting, wonder why an official installation python does not need it. 🤔 this needs to be investigated and determined to see if it's reasonable or not to automatically and when to add it.

> By the way, the new virtualenv seems to install pydoc.bat. The old one didn't - ideally I'd prefer this not to be installed, as it interferes with command completion (myenv\Scripts\py<TAB>). I realise this is a minor use case, but xkcd is on my side 🙂

You can always disable the generation of the Batch activation scripts. pydoc working on other shells and not in Batch was a bug we solved now by adding it.

gaborbernat commented on src/virtualenv/config/convert.py:
```diff
@@ -0,0 +1,69 @@
+    "no": False,
+    "false": False,
+    "off": False,
+}
```
worked in the old way, I'd prefer to keep it working.

gaborbernat commented on src/virtualenv/interpreters/create/cpython/common.py:
```diff
@@ -0,0 +1,136 @@
+        symlink(src, dest)
+        dest_str = str(dest)
+        if not access(dest_str, X_OK):
+            chmod(dest_str, 0o755)  # pragma: no cover
```
we had this in the old and is also what venv uses, so kept it; though seems sensible to migrate to those 👍

gaborbernat replied:
changed

Julian commented on setup.cfg:
```diff
@@ -1,81 +1,108 @@
-              Programming Language :: Python :: 3.6
-              Programming Language :: Python :: 3.7
+classifiers =
+    Development Status :: 3 - Alpha
```
Possibly irrelevant since these do nothing and definitely minor but I think it's odd for a piece of software's development status to be non-monotonic -- I know what you're trying to convey, that the rewrite is risky or whatever so it's back to the starting line (that or this was unintentional) but maybe someone will find it odd that something that was marked mature is now alpha software (or at least that something so crucial is marked alpha)

Julian commented on setup.cfg:
```diff
@@ -1,81 +1,108 @@
+    Programming Language :: Python :: 3.7
+    Programming Language :: Python :: 3.6
+    Programming Language :: Python :: 3.5
+    Programming Language :: Python :: 3.4
```
Probably also minor:

I don't see 3.4 running under CI -- assuming that's intentional (though 3.4 is dead anyhow isn't it?) it's a bit a matter of style as to whether you put it here I guess so maybe it's fine -- just pointing it out to be sure you meant to.

gaborbernat replied:
https://github.com/pypa/virtualenv/issues/1472

Julian commented on src/virtualenv/util/path/_pathlib/via_os_path.py:
```diff
@@ -0,0 +1,115 @@
+from __future__ import absolute_import, unicode_literals
```
Maybe it's worth calling out the reason this exists in a comment? I assume it's https://github.com/mcmtroffaes/pathlib2/issues/56

gaborbernat replied:
It's already there where it's used https://github.com/pypa/virtualenv/blob/rewrite/src/virtualenv/util/path/_pathlib/__init__.py#L33

Julian commented on setup.py:
```diff
@@ -1,16 +1,17 @@
+            """
+             # coding: utf-8
+             from __future__ import unicode_literals
+             __version__ = {version!r}
```
Presumably minor too: do you actually mean for this to change to being unicode on Py2?

gaborbernat replied:
yes 😄

Julian commented on src/virtualenv/activation/__init__.py:
```diff
@@ -0,0 +1,19 @@
+from __future__ import absolute_import, unicode_literals
+
+from .bash import BashActivator
```
If this imports all the objects anyhow maybe the modules they come from should be private?

gaborbernat replied:
good point, I'll have a think about it, as part of the pre-release cleanup 👍

Julian commented on src/virtualenv/activation/powershell/__init__.py:
```diff
@@ -0,0 +1,10 @@
+
+class PowerShellActivator(ViaTemplateActivator):
+    def templates(self):
+        yield Path("activate.ps1")
```
Powershell works on stuff beyond Windows?

asottile replied:
yep, also works on linux -- I tried it out for [aoc 2019](https://github.com/anthonywritescode/aoc2019/blob/master/day10/part2.ps1)

gaborbernat replied:
yes 👍 for the last two years or so

Julian replied:
TIL

Julian commented on src/virtualenv/config/convert.py:
```diff
@@ -0,0 +1,69 @@
+
+import logging
+
+BOOLEAN_STATES = {
```
FWIW distutils.util.strtobool has this funny logic.

gaborbernat replied:
I know, it's not stable across pythons and does not map to this entirely, and given distutils is planned to be deprecated prefer it to have it here

Julian commented on src/virtualenv/config/env_var.py:
```diff
@@ -0,0 +1,27 @@
+    :return:
+    """
+    environ_key = "VIRTUALENV_{}".format(key.upper())
+    if environ_key in os.environ:
```
Probably? it doesn't matter, but there are silent unicode conversions happening here (on Py2), becuase you've used unicode_literals everywhere (personally I recommend never using that, but I assume you have an opinion there). I don't think anything blows up here unless for some silly reason key is non-ASCII I guess, so possibly it doesn't matter, but pointing it out just in case.

gaborbernat replied:
I'll cast the key to string 😄

Julian commented on src/virtualenv/config/ini.py:
```diff
@@ -0,0 +1,68 @@
+
+from .convert import convert
+
+DEFAULT_CONFIG_FILE = get_default_config_dir() / "virtualenv.ini"
```
Is it too late to bikeshed what config format the file is in :)?

gaborbernat replied:
you can make your case, but wanted something that has stdlib support...

Julian replied:
That indeed makes things very limited yeah :)

I think all the cases have been made at this point so I'll refrain unless someone else wants to continue painting the shed :D

gaborbernat commented on src/virtualenv/activation/batch/__init__.py:
```diff
@@ -0,0 +1,16 @@
+
+class BatchActivator(ViaTemplateActivator):
+    @classmethod
+    def supports(cls, interpreter):
```
no idea, but left the door open for it; note some activators also have python version requirements (see xonsh) - so there's definitely some dynamic logic when something should or should not be generated

pfmoore commented:
>>I tried to test with Python 2.7, using a custom Python 2.7 install (installed via Scoop) and virtualenv -p C:\Users\xxx\scoop\apps\python-2.7.17\1.0\Python-2.7.17\python.exe xx. The resulting environment didn't include a copy of python27.dll, and so python.exe failed.

> Interesting, wonder why an official installation python does not need it. 🤔 this needs to be investigated and determined to see if it's reasonable or not to automatically and when to add it.

Maybe because an all-users Python 2 install using the python.org installer puts `python27.dll` in the Windows system directory, so the fact that it doesn't get copied doesn't matter in that case? (I think it does that, as I say no official Python 2 installation to hand at the moment to check this...)

>> By the way, the new virtualenv seems to install pydoc.bat. The old one didn't - ideally I'd prefer this not to be installed, as it interferes with command completion (myenv\Scripts\py). I realise this is a minor use case, but xkcd is on my side 🙂

> You can always disable the generation of the Batch activation scripts. pydoc working on other shells and not in Batch was a bug we solved now by adding it.

Ah, if it's considered a "batch" activation script, then yes I could do this, as I use PowerShell, so the batch scripts are not very useful to me. Thanks for the suggestion.

pfmoore commented:
> with PEP-517/518 its increasingly unnecessary to have this build dependency, as pip would put it into its build env anyway, its however potentially still creating problems for certain legacy packages, as the pip cant build a wheel directly (unchecked, need to check if it holds true)

For some reason I can't reply inline to this comment :-(

Yes, when using the legacy (non-PEP 517) code path, pip will check if `wheel` is importable, and only build a wheel if it is. In the absence of wheel, it will fall back to `setup.py install`, which has a number of differences in semantics, as well as not being cache-able.

pfmoore commented:
> The resulting environment didn't include a copy of python27.dll, and so python.exe failed

Confirmed with a clean install. #1484 raised for this.

gaborbernat commented on setup.cfg:
```diff
@@ -1,81 +1,108 @@
+    fish = virtualenv.activation.fish:FishActivator
+    power-shell = virtualenv.activation.powershell:PowerShellActivator
+    python = virtualenv.activation.python:PythonActivator
+    xonosh = virtualenv.activation.xonosh:XonoshActivator
```
indeed

gaborbernat commented on src/virtualenv/activation/python/activate_this.py:
```diff
@@ -0,0 +1,57 @@
+prev = set(sys.path)
+
+# fmt: off
+# turn formatter off as json dumps will contain " characters - so we really need here ' black
```
good catch; initially did not json dump with triple quotes 👍

gaborbernat commented on src/virtualenv/config/env_var.py:
```diff
@@ -0,0 +1,27 @@
+            as_type = convert(value, as_type, source)
+            return as_type, source
+        except Exception:
+            pass
```
the converter already logs a warning when failures happen

Julian commented on src/virtualenv/interpreters/create/creator.py:
```diff
@@ -0,0 +1,182 @@
+        )
+
+        parser.add_argument(
+            "dest_dir", help="directory to create virtualenv at", type=cls.validate_dest_dir, default="env", nargs="?",
```
One thing that occurred to me today -- after you pointed out this behavior to me -- I think we may consider that folks will want to put the default name that's chosen here in their gitignores under the presumption that that's someone creating a venv (I think for better or worse folks already occasionally do this for directories called `venv`).

Maybe as is it's still fine, but I have at least a suspicion that a slightly more specific name will satisfy that desire a bit more.

Is there a reason you didn't choose to name this `venv` by default, or was it arbitrarily / shorter to use `env`?

gaborbernat replied:
mostly upon the assumption that's shorter; also python 2 virtual environments are not venvs, so I think helps say some-type of environment 🤔

gaborbernat replied:
though maybe venv is better as it's a virtual environment 🤷‍♂ I'll change it to that

gaborbernat commented on src/virtualenv/activation/bash/__init__.py:
```diff
@@ -0,0 +1,14 @@
+    def supports(cls, interpreter):
+        return interpreter.os != "nt"
+
+    def templates(self):
```
I'm not sure I follow here 🤔

gaborbernat commented on src/virtualenv/run.py:
```diff
@@ -0,0 +1,190 @@
+    return session
+
+
+def session_via_cli(args):
```
reorganized under #1487

gaborbernat commented:
I've addressed all points raised inside https://github.com/pypa/virtualenv/pull/1487 (minus three that have their own issues); will close this now, if anyone has further feedback please open a new PR and do it there; or reach out to me. Thanks for everyone for their effort.

Julian commented:
Woohoo! In case it hasn't been said enough yet, @gaborbernat you are great thanks for moving this forward it is awesome.

gaborbernat commented on src/virtualenv/interpreters/create/cpython/cpython2.py:
```diff
@@ -0,0 +1,69 @@
+
+from virtualenv.util.path import Path, copy
+
+from .common import CPython, CPythonPosix, CPythonWindows
```
mostly because I did not find a tool to automate it 😢 any suggestions? I'd prefer to make relative what's possible without (or at most 1 level ``..`` and keep everything else as absolute)

gaborbernat commented on setup.cfg:
```diff
@@ -1,81 +1,108 @@
+virtualenv.activation.batch = *.bat
+virtualenv.activation.fish = *.fish
+virtualenv.activation.powershell = *.ps1
+virtualenv.activation.xonosh = *.xsh
```
fixed already part of the feedback part 👍

gaborbernat commented on MANIFEST.in:
```diff
@@ -1,22 +1,12 @@
-include virtualenv.py
```
that's a good idea 👍

gaborbernat replied:
hmm pushing an orphan branch takes a while 🤔

gaborbernat replied:
orphan branch created under https://github.com/pypa/virtualenv/commits/rewrite

gaborbernat commented on src/virtualenv/error.py:
```diff
@@ -0,0 +1,12 @@
+    """Failed a process call"""
+
+    def __init__(self, code, out, err, cmd):
+        self.code = code
```
why? 🤔

Julian replied:
Calling `super` for exceptions sets `Exception.args` -- lots of people forget that (calling `super`), so personally I never rely on it being set, but in theory it's "nice" -- though personally I was going to comment that `RuntimeError` is to me not a great `Exception` to subclass (I know it's what was here originally though).

asottile replied:
iirc it also affects the pickleability of exceptions -- though that probably doesn't matter here (it's a common bug in python stdlib though if you want some easy cpython patches!)

Julian replied:
>pickleability of exceptions

You're a braver man than I am Anthony.

asottile replied:
\***cries in flake8**\*
